### PR TITLE
enable multistage builds

### PR DESCRIPTION
**Pull request by sosiouxme**
Enhance atomic-reactor to correctly manage multi-stage build inputs and metadata.

Implementation of OSBS-5663. The commits are intended to be individually coherent for easier reviews.

### CODE REVIEW

File: atomic_reactor/build.py [hunk 1]
```diff
@@ -166,13 +168,20 @@ def df_path(self):
...
+        self.set_base_image(base)
         logger.debug("base image specified in dockerfile = '%s'", self.base_image)
-        if not self.base_image.tag:
-            self.base_image.tag = 'latest'
+        self.parent_images.clear()
+        for image in dfp.parent_images:
+            self.parent_images[image] = None
 
     def set_base_image(self, base_image):
+        logger.info("setting base image to '%s'", base_image)
```
Comment (mlangsdorf) [inline atomic_reactor/build.py]: Is this really necessary? We're already pretty noisy.

File: atomic_reactor/build.py [hunk 2]
```diff
@@ -166,13 +168,20 @@ def df_path(self):
...
+        base = dfp.baseimage
+        if base is None:
+            raise RuntimeError("no base image specified in Dockerfile")
+        self.set_base_image(base)
         logger.debug("base image specified in dockerfile = '%s'", self.base_image)
-        if not self.base_image.tag:
-            self.base_image.tag = 'latest'
+        self.parent_images.clear()
+        for image in dfp.parent_images:
+            self.parent_images[image] = None
```
Comment (mlangsdorf) [inline atomic_reactor/build.py]: when is self.parent_images ever referenced? you set it here, but AFAIK never use it again.

File: atomic_reactor/plugins/pre_pull_base_image.py [hunk 3]
```diff
@@ -65,92 +87,78 @@ def resolve_base_image(self, build_json):
...
-            self.log.info("'%s' not found", base_image_with_registry.to_str())
-            base_image_with_registry.namespace = 'library'
-            self.log.info("trying '%s'", base_image_with_registry.to_str())
+        return image_with_registry
 
+    def _pull_and_tag_image(self, image, build_json):
+        """Docker pull the image and tag it uniquely for use by this build"""
+        image = image.copy()
+        first_library_exc = None
+        for _ in range(20):
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/pre_pull_base_image.py]: Please add documentation for the logic here.

File: atomic_reactor/plugins/pre_pull_base_image.py [hunk 4]
```diff
@@ -65,92 +87,78 @@ def resolve_base_image(self, build_json):
...
-        # Attempt to tag it using a unique ID. We might have to retry
-        # if another build with the same parent image is finishing up
-        # and removing images it pulled.
-
-        # Use the OpenShift build name as the unique ID
-        unique_id = build_json['metadata']['name']
-        original_base_image = base_image.copy()
-        base_image = ImageName(repo=unique_id)
-
-        for _ in range(20):
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/pre_pull_base_image.py]: any reason why we try this 20 times, as opposed to 10 or 5 or 100? I get that you're inheriting this code, mostly curious.

File: tests/plugins/test_change_from_in_df.py [hunk 5]
```diff
@@ -14,80 +14,171 @@
...
     dfp = df_parser(str(tmpdir))
     dfp.content = df_content
 
-    workflow = mock_workflow(dfp.dockerfile_path, dict(no_id="here"))
-    result = runner(docker_tasker, workflow).run()
+    workflow = mock_workflow(dfp.dockerfile_path)
+    workflow.builder.tasker.inspect_image = lambda img: dict(no_id="here")
+
+    with pytest.raises(NoIdInspection):
+        ChangeFromPlugin(docker_tasker(), workflow).run()
```
Comment (mlangsdorf) [inline tests/plugins/test_change_from_in_df.py]: run_plugin could take a second optional argument of "should_raise" that defaults to None and do pytest.raises(should_raise). I suggest that because after nicely encapsulating ChangeFromPlugin in your run_plugin() function, it's jarring to see it again being called directly in the tests cases.

File: atomic_reactor/build.py (hunk 2 above)
Comment (sosiouxme) [reply atomic_reactor/build.py]: it's filled in later in pull_base_image to record what local tag or digest is used for those images -- used to update the dockerfile FROMs and recorded in the build metadata

File: atomic_reactor/build.py (hunk 1 above)
Comment (sosiouxme) [reply atomic_reactor/build.py]: Not really, just was helpful in some of my later debugging of integration tests. To me, verbosity is good as logs are about all we have for debugging builds, and nobody looks at them unless something goes wrong. But I can take this out if you like :)

File: atomic_reactor/plugins/pre_pull_base_image.py (hunk 4 above)
Comment (sosiouxme) [reply atomic_reactor/plugins/pre_pull_base_image.py]: I don't know, of course. You only go through the loop if the image pulls successfully but gets immediately deleted before being tagged. That ought to be an edge case race condition, so 20 seems excessive, but on the other hand there could be a *lot* of builds running with the same parents, and retries are cheap.

File: tests/plugins/test_change_from_in_df.py (hunk 5 above)
Comment (sosiouxme) [reply tests/plugins/test_change_from_in_df.py]: There are a number of situations where it makes more sense to go straight to the class to test a particular behavior, and there's a very specific reason for it here.

I like to raise named exceptions so I can test for exactly the thing that is expected to go wrong without relying on the text. But if you import the exception class from the plugin, and then run the plugin via a plugin-runner, the comparison of the exception you got to the expected class will fail. This is because the plugin runner reloads all the modules before running the plugin and what you get back is technically a different class, even though it came from the same code. So the level of indirection really gets in the way here.

File: atomic_reactor/plugins/pre_pull_base_image.py (hunk 3 above)
Comment (sosiouxme) [reply atomic_reactor/plugins/pre_pull_base_image.py]: I'm all for docs. Are you talking about comments in the code, or docs in the repo, or on readthedocs....?

And "the logic here" being the behavior of applying a unique tag to the parent image for the duration of the build?

Comment (mlangsdorf) [reply atomic_reactor/plugins/pre_pull_base_image.py]: inline docs of whatever is going on in this for loop, please.

File: atomic_reactor/build.py (hunk 1 above)
Comment (mlangsdorf) [reply atomic_reactor/build.py]: I'm not fond of it but I accept your justification.

File: atomic_reactor/build.py (hunk 2 above)
Comment (mlangsdorf) [reply atomic_reactor/build.py]: Okay. You might want to add a note about that in the commit message.

**Comment by mlangsdorf** [comment]
Minor nitpicks and questions but generally looks good. I would like to see the changes to test_change_from_in_df squashed to the original creation patch but no other major changes requested.

File: atomic_reactor/build.py (hunk 2 above)
Comment (sosiouxme) [reply atomic_reactor/build.py]: Will do!

File: atomic_reactor/plugins/pre_pull_base_image.py (hunk 4 above)
Comment (mlangsdorf) [reply atomic_reactor/plugins/pre_pull_base_image.py]: Yes, thanks.

**Comment by sosiouxme** [comment]
poke @rcerven @twaugh @lcarva

File: atomic_reactor/plugins/pre_pull_base_image.py [hunk 6]
```diff
@@ -65,15 +65,22 @@ def run(self):
         Pull parent images and retag them uniquely for this build.
         """
         build_json = get_build_json()
-        base_image = self._resolve_base_image(build_json)
-        base_image_with_registry = self._ensure_image_registry(base_image)
+        base_image_str = str(self.workflow.builder.base_image)
+        for nonce, parent in enumerate(sorted(self.workflow.builder.parent_images.keys())):
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_pull_base_image.py]: Might `nonce` contain ":" here? If so, won't using that value as an image tag fail?

File: atomic_reactor/plugins/pre_change_from_in_df.py [hunk 7]
```diff
@@ -29,21 +39,54 @@ def __init__(self, tasker, workflow):
...
-                sys.stdout.write(line)
+        builder = self.workflow.builder
+        dfp = DockerfileParser(builder.df_path)
+
+        df_base = dfp.baseimage
+        build_base = builder.base_image.to_str()
+        if build_base != df_base and build_base != builder.parent_images[df_base]:
+            # something updated the build's base_image without updating the corresponding
+            # parent_images entry; that should be OK, let base_image override.
+            builder.parent_images[df_base] = build_base
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_change_from_in_df.py]: Should `builder.base_image` be a property which handles updating `self.parent_images`?

Comment (sosiouxme) [reply atomic_reactor/plugins/pre_change_from_in_df.py]: I debated that a few times and ultimately decided not to. However I'm reconsidering now because it rested on not having ready access everywhere to know which key in `parent_images` was from the Dockerfile baseimage. Since later I made `set_base_image` also set `original_base_image` with that info, that now seems like a lame excuse for not keeping the two in sync automatically.

File: atomic_reactor/util.py [hunk 8]
```diff
@@ -1183,6 +1186,25 @@ def get_primary_images(workflow):
     return primary_images
 
 
+def get_parent_image_koji_data(workflow):
+    """Transform koji_parent plugin results into metadata dict."""
+    koji_parent = workflow.prebuild_results.get(PLUGIN_KOJI_PARENT_KEY) or {}
+
+    image_metadata = {}
+    image_metadata[PARENT_IMAGE_BUILDS_KEY] = koji_parent.get(PARENT_IMAGES_KOJI_BUILDS) or {}
```
Comment (lcarva) [inline atomic_reactor/util.py]: Let's be sure to document this additional piece of metadata.

File: tests/plugins/test_change_from_in_df.py [hunk 9]
```diff
@@ -14,80 +14,173 @@
...
-    if MOCK:
-        mock_docker()
 
+def mock_workflow(df_path, base_image="mock:default"):
+    """
+    Provide just enough structure that workflow can be used to run a plugin.
+    Defaults below are solely to enable that purpose; tests where those values
+    matter should provide their own.
+    """
     builder = flexmock(
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply tests/plugins/test_change_from_in_df.py]: Probably - I missed that stubs had been committed.

File: atomic_reactor/plugins/pre_pull_base_image.py [hunk 10]
```diff
@@ -65,92 +87,86 @@ def resolve_base_image(self, build_json):
...
+            # if registry specified in Dockerfile image, ensure it's the one allowed by config
+            if image.registry and image.registry != self.parent_registry:
+                error = (
+                    "Registry specified in dockerfile image doesn't match configured one. "
+                    "Dockerfile: '%s'; expected registry: '%s'"
+                    % (image, self.parent_registry))
+                self.log.error("%s", error)
+                raise RuntimeError(error)
+
+            self.log.info(
```
[the remark this answers is not in the document]
Comment (lcarva) [reply atomic_reactor/plugins/pre_pull_base_image.py]: Yeah, probably a good idea.

File: atomic_reactor/plugins/pre_pull_base_image.py [hunk 11]
```diff
@@ -83,6 +90,7 @@ def _resolve_base_image(self, build_json):
         except (TypeError, KeyError, IndexError):
             base_image = self.workflow.builder.base_image
         else:
+            self.log.info("rebuild triggered by image %s; using that as base image.", image_id)
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply atomic_reactor/plugins/pre_pull_base_image.py]: I see, I think... I was just trying to log some kind of justification as to why we're substituting this image... trying to think of something clear and correct. Is it maybe better to say nothing? The base image change will be logged regardless.

[the remark this answers is not in the document]
Comment (lcarva) [reply atomic_reactor/plugins/pre_pull_base_image.py]: I think it's worth mentioning that there was something in triggerBy for debugging purposes.

**Comment by sosiouxme** [comment]
Addressed comments as extra commits - will squash if no one raises issues with them and tests pass.

**Comment by sosiouxme** [comment]
Rebased, squashed, integration tests pass (modulo test issues). Should be ready for review again.

**Comment by sosiouxme** [comment]
poke @twaugh @lcarva @rcerven for final review...

**Comment by twaugh** [comment]
Please update docs/koji.md to reflect the new keys in 'extra'.

**Comment by sosiouxme** [comment]
Done!

**Comment by sosiouxme** [comment]
test.sh installs dockerfile-parse from source. travis sets up test environment and tries to install it from pypi, where 0.0.11 isn't published yet.

For now, I backed up the version in requirements.txt to 0.0.10 in the hopes of not breaking travis.

**Comment by sosiouxme** [comment]
https://travis-ci.org/projectatomic/atomic-reactor/jobs/401014084 reveals travis is also checking the spec file and fails because dockerfile-parse 0.0.11 hasn't been published as a Fedora RPM.

At this point these reqs are likely wildly inaccurate anyway. I'm going to file an issue to publish dockerfile-parse 0.0.11 and then update the atomic-reactor reqs.

**Comment by sosiouxme** [comment]
updated release notes with:

- correctly manage multi-stage build inputs and metadata
